Implement simple Subscription Caching mechanism using UserDefaults

afterxleep opened the pull request:
**Required**:

Task/Issue URL:  https://app.asana.com/0/1204099484721401/1206574474409573/f
iOS PR:  https://github.com/duckduckgo/iOS/pull/2556
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2359
What kind of version bump will this require?: Minor

**Description**:
- Implements mechanism for caching subscription info
- Caches entitlements
- Cleans cache on signout.
- Returns cached entitlements by default (if present)
- Adds a skipCache parameter to `fetchEntitlements` to force fetching from the service
- Posts a Notification only when entitlements change
- Existing API remains the same.  This does not require changes on local implementation

<!--
Tagging instructions
If this PR isn't ready to be merged for whatever reason it should be marked with the `DO NOT MERGE` label (particularly if it's a draft)
If it's pending Product Review/PFR, please add the `Pending Product Review` label.

If at any point it isn't actively being worked on/ready for review/otherwise moving forward (besides the above PR/PFR exception) strongly consider closing it (or not opening it in the first place). If you decide not to close it, make sure it's labelled to make it clear the PRs state and comment with more information.
-->

**Steps to test this PR**:
1.  Purchase a subscription
2. Confirm that entitlements are returned from cache after first use (Breakpoint [here](https://github.com/duckduckgo/BrowserServicesKit/pull/703/files#diff-6024bf9848884f7e6cbb4b48816ebecb139706dd2ee31fb26d08242020aff7f5R268))

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

miasma13 commented on Sources/Subscription/AccountManager.swift:
```diff
@@ -22,6 +22,7 @@ import Common
 public extension Notification.Name {
     static let accountDidSignIn = Notification.Name("com.duckduckgo.subscription.AccountDidSignIn")
     static let accountDidSignOut = Notification.Name("com.duckduckgo.subscription.AccountDidSignOut")
+    static let entitlementsUpdated = Notification.Name("com.duckduckgo.subscription.EntitlementsUpdated")
```
What about `entitlementsDidChange`?

quanganhdo replied:
Second this 👍

miasma13 commented on Sources/Subscription/AccountManager.swift:
```diff
@@ -221,19 +230,44 @@ public class AccountManager: AccountManaging {
-            return .success(entitlements)
+                if entitlements != cachedEntitlements {
+                    subscriptionCache.set(entitlements, forKey: SubscriptionCache.Component.entitlements.rawValue)
+                    NotificationCenter.default.post(name: .entitlementsUpdated, object: self, userInfo: nil)
```
I think it would be handy to also send in the notification the updated set of entitlements.

quanganhdo commented on Sources/Subscription/AccountManager.swift:
```diff
@@ -221,19 +230,44 @@ public class AccountManager: AccountManaging {
+            }
+        }
+
+        public func fetchEntitlements(skipCache: Bool = false) async -> Result<[Entitlement], Error> {
```
Maybe I’m overthinking this, but what if instead of `skipCache` we do something similar to `NSURLRequest.CachePolicy`? E.g. supporting these scenarios based on what the API user wants:

- `reloadIgnoringLocalCacheData`: Same as `skipCache`
- `returnCacheDataElseLoad`: Use existing cache, load from remote only if there’s no cached data
- `returnCacheDataDontLoad`: Use existing cache, fails if no cached data is available

I can see the entitlement monitoring code that NetP has uses `returnCacheDataElseLoad` behavior.

quanganhdo commented on Sources/Subscription/AccountManager.swift:
```diff
@@ -221,19 +230,44 @@ public class AccountManager: AccountManaging {
+                return .success(entitlements)
+
+            case .failure(let error):
+                subscriptionCache.cleanup()
```
Won’t this accidentally clear the cache when there’s a connection or temporary server error?

afterxleep commented:
@miasma13 @quanganhdo got some updates...

1. Made the cache generic so we can not initialize for a specific type
2. Implemented Apple-like cache policy methods for elegance :bowtie: 
3. Removed cache cleanup on fetch failure
4. Added entitlements to notification object
5. Addressed small comments for vars and etc
